proposals/0631-main-return-types.rst

```diff
@@ -0,0 +1,214 @@
+ -- but users might expect 'Termination' to be the class of
+ -- /allowed/ main return types (and perhaps some day it
+ -- could become that).
+ instance Termination () where
+   report = const ExitSuccess
+
+ instance Termination Void where
+   report = absurd
+
+Eventually, ``Termination`` and its instances should move into ``base``,
+probably in ``System.Exit``.
+
+Examples
+--------
+
+::
+
+ data ExitReason
+   = AllsWell
+   | NoResults
+
+ instance Termination ExitReason where
+   report AllsWell = ExitSuccess
+   report NoResults = ExitFailure 1
+
+ main = do
+   results <- doSomeWork
+   case results of
+     [] -> pure NoResults
+     _ -> print results >> pure AllsWell
+
+With this proposal in place, the program would exit with code ``1`` if
+``doSomeWork`` doesn't yield results.
+
+Effect and Interactions
+-----------------------
+This change will allow users to write ``main`` programs which yield an
+``ExitCode`` and have that code respected.
+
+No known interactions.
+
+Costs and Drawbacks
+-------------------
+Development seems likely to be minimal, and maintenance basically
+non-existent. This will have no *negative* impact on novice users
+learning the language.
+
+
+Backward Compatibility
+----------------------
+Technically, this will cause a change in behavior in extremely rare cases
+(this is level ``1`` on the breakage scale): If a program currently has
+``main :: IO ExitCode`` and some branch currently yields ``ExitFailure n``,
+the program will currently exit with ``0`` on that branch. After this proposal,
+it will exit with ``n``. But the latter is almost certainly the desired behavior for
+writing code like that, and I'd be surprised if it exists in any real program.
+
+Because this behavior change, if it ever actually matters, is likely in the direction
+of *improving* the program behavior, it's not clear that there are any real costs
+to this "breakage". Still, we could add a warning to the ``Termination ExitCode``
+instance and encourage users to use a custom type, or see the alternatives
+for `an option <#no-exitcode-instance>`_ with no backwards incompatibility.
+
+Alternatives
+------------
+
+Retain the status quo
+^^^^^^^^^^^^^^^^^^^^^
+
+While the current behavior is functional and any surprise is likely to be
+caught early in testing, this change will make learning Haskell a bit
+easier and arguably makes for ``main`` functions which better match
+the Haskell ethos of well-typed structured interfaces.
+
+If we did stick to the status quo, users could perhaps
+have ``mainWithTermination :: Termination a => IO a -> IO ()``
+and always define ``main = mainWithTermination $ do { ... }``.
+
+Do it just for ExitCode
+^^^^^^^^^^^^^^^^^^^^^^^
+
+Rather than a typeclass, we could just have ``main :: IO ExitCode`` be
+respected. Using the typeclass allows using domain-specific types
+to exit codes as in `the example <#Examples>`_, letting the user
+give semantic meaning to the exit status in the typical Haskell
+way.
```

Comment:
I don't understand what this alternative means.

Reply:
Fixed.